include/lock/pessimistic_lock.hpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2021 Database Group, Nagoya University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CPP_UTILITY_PESSIMISTIC_LOCK_HPP
+#define CPP_UTILITY_PESSIMISTIC_LOCK_HPP
+
+#ifndef SPINLOCK_HINT
+#ifdef CPP_UTILITY_HAS_SPINLOCK_HINT
+#include <xmmintrin.h>
+#define SPINLOCK_HINT _mm_pause();  // NOLINT
+#else
+#define SPINLOCK_HINT /* do nothing */
+#endif
+#endif
+
+#include <atomic>
+
+namespace dbgroup::lock
+{
+class PessimisticLock
+{
+ public:
+  /*####################################################################################
+   * Public constructors and assignment operators
+   *##################################################################################*/
+
+  constexpr PessimisticLock() = default;
+
+  PessimisticLock(const PessimisticLock&) = delete;
+  PessimisticLock(PessimisticLock&&) = delete;
+
+  auto operator=(const PessimisticLock&) -> PessimisticLock& = delete;
+  auto operator=(PessimisticLock&&) -> PessimisticLock& = delete;
+
+  /*####################################################################################
+   * Public destructors
+   *##################################################################################*/
+
+  ~PessimisticLock() = default;
+
+  /*####################################################################################
+   * Public getters
+   *##################################################################################*/
+
+  void
+  LockShared()
+  {
+    auto expected = lock_.load(std::memory_order_acquire) & (~1UL);  // turn-off LSB
+    auto desired = expected + 2;                                     // increment read-counter
+    while (!lock_.compare_exchange_strong(expected, desired, std::memory_order_relaxed,
+                                          std::memory_order_acquire)) {
```

Comment:
見逃してましたが，リトライが発生しうるCASは基本`weak`でよいです．逆に，リトライなしの1回だけのときに`strong`を使います．

あと，他のところを直してたらここはこっちのフェンスの方がよい気がしてきました．比較が成功したときにだけ（i.e., ロックが取得できたときだけ）フェンスをはれれば十分なので．

```suggestion
    auto expected = lock_.load(std::memory_order_relaxed) & (~1UL);  // turn-off LSB
    auto desired = expected + 2;                                     // increment read-counter
    while (!lock_.compare_exchange_weak(expected, desired, std::memory_order_acquire,
                                          std::memory_order_relaxed)) {
```